plugins/powerpc/powerpc_branch.ml

```diff
@@ -0,0 +1,277 @@
+  let sh = unsigned of_string "0b00" in
+  let ctr_ok = unsigned var bit in
+  let cond_ok = unsigned var bit in
+  let x = unsigned var (bitwidth 5) in
+  let tm = signed var halfword in
+  RTL.[
+    x := last bo 5;
+    when_ (nth bit x 2 = zero) [
+      cpu.ctr := cpu.ctr - one;
+    ];
+    ctr_ok := nth bit x 2 lor ((cpu.ctr <> zero) lxor (nth bit x 3));
+    cond_ok := nth bit x 0 lor (bi lxor (lnot (nth bit x 1)));
+    when_ (ctr_ok land cond_ok) [
+      tm := bd ^ sh;
+      cpu.jmp (cpu.pc + tm);
+    ]
+  ]
+
+let bca cpu ops =
+  let bo = unsigned imm ops.(0) in
+  let bi = unsigned cpu.reg ops.(1) in
+  let bd = unsigned imm ops.(2) in
+  let sh = unsigned of_string "0b00" in
+  let ctr_ok = unsigned var bit in
+  let cond_ok = unsigned var bit in
+  let x = unsigned var (bitwidth 5) in
+  let tm = signed var cpu.word_width in
+  RTL.[
+    x := last bo 5;
+    when_ (nth bit x 2 = zero) [
+      cpu.ctr := cpu.ctr - one;
+    ];
+    ctr_ok := nth bit x 2 lor ((cpu.ctr <> zero) lxor (nth bit x 3));
+    cond_ok := nth bit x 0 lor (bi lxor (lnot (nth bit x 1)));
+    when_ (ctr_ok land cond_ok) [
+      tm := bd ^ sh;
+      cpu.jmp tm;
+    ]
+  ]
+
+let bcl = bc ^ update_link_register
+let bcla = bca ^ update_link_register
+
+(** bdz  target = bc 18,0, target *)
+let bdz cpu ops =
+  let bd = unsigned imm ops.(0) in
+  let sh = unsigned of_string "0b00" in
+  let tm = signed var cpu.word_width in
+  RTL.[
+    cpu.ctr := cpu.ctr - one;
+    when_ (low cpu.word_width cpu.ctr = zero) [
+      tm := bd ^ sh;
+      cpu.jmp (cpu.pc + tm)
+    ]
+  ]
+
+(** bdnz  target = bc 16,0, target *)
+let bdnz cpu ops =
+  let bd = unsigned imm ops.(0) in
+  let sh = unsigned of_string "0b00" in
+  let tm = signed var cpu.word_width in
+  RTL.[
+    cpu.ctr := cpu.ctr - one;
+    when_ (low cpu.word_width cpu.ctr <> zero) [
+      tm := bd ^ sh;
+      cpu.jmp (cpu.pc + tm)
+    ]
+  ]
+
+(** Branch Instructions, Branch Conditional to Link Cpu.Register
+    Page 38 of IBM Power ISATM Version 3.0 B
+    examples:
+    4e 9f 00 20 	bclr	 20, 31
+    4e 9f 00 21 	bclrl	 20, 31 *)
+let bclr cpu ops =
+  let bo = unsigned imm ops.(0) in
+  let bi = unsigned cpu.reg ops.(1) in
+  let sh = unsigned of_string "0b00" in
+  let ctr_ok = unsigned var bit in
+  let cond_ok = unsigned var bit in
+  let x = unsigned var (bitwidth 5) in
+  RTL.[
+    x := last bo 5;
+    when_ (nth bit x 2 = zero) [
+      cpu.ctr := cpu.ctr - one;
+    ];
+    ctr_ok := nth bit x 2 lor ((cpu.ctr <> zero) lxor (nth bit x 3));
+    cond_ok := nth bit x 0 lor (bi lxor (lnot (nth bit x 1)));
+    when_ (ctr_ok land cond_ok) [
+      cpu.jmp (cpu.lr ^ sh);
+    ];
+  ]
+
+let bclrl = bclr ^ update_link_register
+
+(** Branch Instructions extended mnemonic, branch to LR unconditionally.
+    Page 792 of IBM Power ISATM Version 3.0 B
+    examples:
+    4e 80 00 20   blr
+    4e 80 00 21   blrl *)
+let blr cpu ops =
+  let sh = unsigned of_string "0b00" in
```

Comment:
I don't like this implementation. We shall try to follow the actual CPU microcode as close as possible. I understand that this is how it is written in the manual, that an word is concatenated with two zero bits. But it would be much nicer (and closer to what actually happens in the CPU) if we will just use the shift left operation.

Also, I don't think it is a good idea to use the string representation of a binary word here. Not sure whether it is worthwhile anywhere.